[Keymap] Add ollyhayes personal keymap

ollyhayes opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

Add my keymap for the crkbd keyboard

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

drashna commented on keyboards/crkbd/keymaps/ollyhayes/README.md:
```diff
@@ -0,0 +1,67 @@
+# ollyhayes's Keymap for [Corne Keyboard (crkbd)](https://github.com/foostan/crkbd)
```
please rename this to be lower case.

drashna commented on keyboards/crkbd/keymaps/ollyhayes/config.h:
```diff
@@ -0,0 +1,107 @@
+
+#define EE_HANDS
+
+// #define USE_SERIAL_PD2
```
```suggestion
```

drashna commented on keyboards/crkbd/keymaps/ollyhayes/config.h:
```diff
@@ -0,0 +1,107 @@
+#define TAPPING_TERM 100
+
+#define NO_ACTION_ONESHOT
+#define NO_ACTION_MACRO
+#define NO_ACTION_FUNCTION
```
```suggestion
```

drashna commented on keyboards/crkbd/keymaps/ollyhayes/config.h:
```diff
@@ -0,0 +1,107 @@
+
+#ifdef RGBLIGHT_ENABLE
+    #undef RGBLED_NUM
+    #define RGBLIGHT_ANIMATIONS
```
Could you explicitly define which animations are enabled? 
```suggestion
#define RGBLIGHT_EFFECT_BREATHING
#define RGBLIGHT_EFFECT_RAINBOW_MOOD
#define RGBLIGHT_EFFECT_RAINBOW_SWIRL
#define RGBLIGHT_EFFECT_SNAKE
#define RGBLIGHT_EFFECT_KNIGHT
#define RGBLIGHT_EFFECT_CHRISTMAS
#define RGBLIGHT_EFFECT_STATIC_GRADIENT
#define RGBLIGHT_EFFECT_RGB_TEST
#define RGBLIGHT_EFFECT_ALTERNATING
#define RGBLIGHT_EFFECT_TWINKLE
```

drashna commented on keyboards/crkbd/keymaps/ollyhayes/config.h:
```diff
@@ -0,0 +1,107 @@
+#   define RGB_MATRIX_KEYPRESSES // reacts to keypresses
+// #   define RGB_MATRIX_KEYRELEASES // reacts to keyreleases (instead of keypresses)
+// #   define RGB_DISABLE_AFTER_TIMEOUT 0 // number of ticks to wait until disabling effects
+#   define RGB_DISABLE_WHEN_USB_SUSPENDED true // turn off effects when suspended
```
```suggestion
#   define RGB_DISABLE_WHEN_USB_SUSPENDED // turn off effects when suspended
```

drashna commented on keyboards/crkbd/keymaps/ollyhayes/keymap.c:
```diff
@@ -0,0 +1,116 @@
+#include "ollyhayes.h"
+
+#ifdef OLED_ENABLE
+#    include "oled.c"
```
Please don't include c files.  You should use rules.mk to add it (`SRC += oled.c`).  Header files should be included, only.

drashna commented on keyboards/crkbd/keymaps/ollyhayes/oled.c:
```diff
@@ -0,0 +1,119 @@
+}
+
+void oled_render_rgb_mode(void) {
+    static char mode_str[6] = {};
+    snprintf(
+        mode_str,
+        sizeof(mode_str),
+        "%u",
+        rgb_matrix_get_mode());
```
```suggestion
```

drashna commented on keyboards/crkbd/keymaps/ollyhayes/oled.c:
```diff
@@ -0,0 +1,119 @@
+        rgb_matrix_get_mode());
+
+    oled_write_P(PSTR("Animation mode: "), false);
+    oled_write_ln(mode_str, false);
```
```suggestion
    oled_write_ln(get_u8_str(rgb_matrix_get_mode(), ' '), false);
```

drashna commented on keyboards/crkbd/keymaps/ollyhayes/oled.c:
```diff
@@ -0,0 +1,119 @@
+        ticks++;
+    }
+
+    snprintf(
+        ticks_str,
+        sizeof(ticks_str),
+        "%u",
+        ticks);
```
```suggestion
```

drashna commented on keyboards/crkbd/keymaps/ollyhayes/oled.c:
```diff
@@ -0,0 +1,119 @@
+void oled_render_timer(void) {
+    static uint16_t microticks = 0;
+    static uint16_t ticks = 0;
+    static char ticks_str[6] = {};
```
```suggestion
```

drashna commented on keyboards/crkbd/keymaps/ollyhayes/oled.c:
```diff
@@ -0,0 +1,119 @@
+        ticks);
+
+    oled_write_P(PSTR("Ticks: "), false);
+    oled_write_ln(ticks_str, false);
```
```suggestion
    oled_write_ln(get_u16_str(ticks, ' '), false);
```

drashna commented on keyboards/crkbd/keymaps/ollyhayes/oled.c:
```diff
@@ -0,0 +1,119 @@
+}
+
+void oled_render_presses(void) {
+    static char presses_str[6] = {};
+    snprintf(
+        presses_str,
+        sizeof(presses_str),
+        "%u",
+        key_presses);
```
```suggestion
```

drashna commented on keyboards/crkbd/keymaps/ollyhayes/oled.c:
```diff
@@ -0,0 +1,119 @@
+        key_presses);
+
+    oled_write_P(PSTR("Presses: "), false);
+    oled_write_ln(presses_str, false);
```
```suggestion
    oled_write_ln(get_u16_str(presses_str, ' '), false);
```

drashna commented on users/ollyhayes/ollyhayes.c:
```diff
@@ -0,0 +1,203 @@
+
+#include "ollyhayes.h"
+
+#include <stdio.h>
```
```suggestion
```

drashna commented on users/ollyhayes/ollyhayes.c:
```diff
@@ -0,0 +1,203 @@
+
+static void sendUtf(const char *str) {
+    // sequence for linux
+    register_code(KC_LCTRL);
+    register_code(KC_LSFT);
+    tap_code(KC_U);
+    unregister_code(KC_LSFT);
+    unregister_code(KC_LCTRL);
```
```suggestion
    tap_code16(C(S(KC_U)));
```

drashna commented on users/ollyhayes/ollyhayes.c:
```diff
@@ -0,0 +1,203 @@
+}
+
+layer_state_t default_layer_state_set_kb(layer_state_t state) {
+    if (state == 1UL<<BASE) {
```
```suggestion
    if (layer_state_cmp(state, BASE)) {
```

drashna commented on users/ollyhayes/ollyhayes.c:
```diff
@@ -0,0 +1,203 @@
+layer_state_t default_layer_state_set_kb(layer_state_t state) {
+    if (state == 1UL<<BASE) {
+        rgb_matrix_mode(3);
+    } else if (state == 1UL<<GAMES) {
```
```suggestion
    } else if (layer_state_cmp(state,GAMES)) {
```

drashna commented on users/ollyhayes/ollyhayes.c:
```diff
@@ -0,0 +1,203 @@
+
+        case CLEFT:
+            if (record->event.pressed) {
+                register_code(KC_LCTRL);
+                tap_code(KC_LEFT);
+                unregister_code(KC_LCTRL);
```
`C(KC_LEFT)`

drashna commented on users/ollyhayes/ollyhayes.c:
```diff
@@ -0,0 +1,203 @@
+
+        case CRIGHT:
+            if (record->event.pressed) {
+                register_code(KC_LCTRL);
+                tap_code(KC_RIGHT);
+                unregister_code(KC_LCTRL);
```
`C(KC_RIGHT)`

drashna commented on users/ollyhayes/ollyhayes.c:
```diff
@@ -0,0 +1,203 @@
+#   include "virtser.h"
+#endif
+
+static void sendUtf(const char *str) {
```
Why not just use unicode?  You're essentially duplicating the code here, anyways. 

https://docs.qmk.fm/#/feature_unicode

drashna replied:
Specifically, you can use `send_unicode_string(str)` for this instead.  You'd want to enable unicode, and you could use `UNICODE_COMMON = yes` in your rules, and just use the core unicode framework.  You'd just need to set the mode.

ollyhayes replied:
Thanks! I switched to using the unicode feature and added the commit to the review.

ollyhayes commented:
@drashna, I think I've addressed all your comments, and also added a few other things, and I rebased against the latest upstream/master. I've left them as separate commits so you can see what I did, shall I squash them all together?

I just have one other question, I never managed to get my 12 underglow leds to solder properly so I just left them out and shorted the data line, and then I changed `g_led_config`  in my branch to match the new set of leds like this: https://github.com/ollyhayes/qmk_firmware/commit/525cfdec3af10bd7ea793d65dcde8cb5afdc55dd

Obviously I can't submit this commit as it is, unless there an easy way to redefine this variable only for my keymap? If not I will just leave this commit only in my local repository.

fauxpark commented on keyboards/crkbd/keymaps/ollyhayes/keymap.c:
```diff
@@ -0,0 +1,112 @@
+#include "ollyhayes.h"
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+  [BASE] = LAYOUT_split_3x6_3( \
+// +------------+------+------+----------+----------+---------+--------+----------+----------+--------+---------+---------+
+     NUM_SWITCH , KC_Q , KC_W , KC_F     , KC_P     , KC_G    , KC_J   , KC_L     , KC_U     , KC_Y   , KC_SCLN , KC_BSPC , \
+     KC_ESC     , KC_A , KC_R , KC_S     , KC_T     , KC_D    , KC_H   , KC_N     , KC_E     , KC_I   , KC_O    , KC_ENT  , \
+     KC_LSFT    , KC_Z , KC_X , KC_C     , KC_V     , KC_B    , KC_K   , KC_M     , KC_COMM  , KC_DOT , KC_SLSH , KC_RSFT , \
+                                KC_LCTRL , MO(PUNC) , MO(ARR) , KC_SPC , MO(PUNC) , KC_RCTRL                                \
```
```suggestion
  [BASE] = LAYOUT_split_3x6_3(
// +------------+------+------+----------+----------+---------+--------+----------+----------+--------+---------+---------+
     NUM_SWITCH , KC_Q , KC_W , KC_F     , KC_P     , KC_G    , KC_J   , KC_L     , KC_U     , KC_Y   , KC_SCLN , KC_BSPC ,
     KC_ESC     , KC_A , KC_R , KC_S     , KC_T     , KC_D    , KC_H   , KC_N     , KC_E     , KC_I   , KC_O    , KC_ENT  ,
     KC_LSFT    , KC_Z , KC_X , KC_C     , KC_V     , KC_B    , KC_K   , KC_M     , KC_COMM  , KC_DOT , KC_SLSH , KC_RSFT ,
                                KC_LCTL , MO(PUNC) , MO(ARR) , KC_SPC , MO(PUNC) , KC_RCTL
```

fauxpark commented on keyboards/crkbd/keymaps/ollyhayes/keymap.c:
```diff
@@ -0,0 +1,112 @@
+// +---------+------+------+---------+---------+---------+---------+---------+---------+--------+---------+---------+
+  ),
+
+  [GAMES] = LAYOUT_split_3x6_3( \
+//  +---------+------+------+---------+---------+----------+---------+---------+---------+--------+---------+---------+
+	  KC_TRNS , KC_Q , KC_W , KC_E    , KC_R    , KC_T     , KC_Y    , KC_U    , KC_I    , KC_O   , KC_P    , KC_TRNS , \
+	  KC_TRNS , KC_A , KC_S , KC_D    , KC_F    , KC_G     , KC_H    , KC_J    , KC_K    , KC_L   , KC_SCLN , KC_TRNS , \
+	  KC_TRNS , KC_Z , KC_X , KC_C    , KC_V    , KC_B     , KC_N    , KC_M    , KC_COMM , KC_DOT , KC_SLSH , KC_TRNS , \
+							  KC_TRNS , KC_LALT , KC_SPACE , MO(ARR) , KC_TRNS , KC_TRNS                                \
```
```suggestion
  [GAMES] = LAYOUT_split_3x6_3(
//  +---------+------+------+---------+---------+----------+---------+---------+---------+--------+---------+---------+
	  KC_TRNS , KC_Q , KC_W , KC_E    , KC_R    , KC_T     , KC_Y    , KC_U    , KC_I    , KC_O   , KC_P    , KC_TRNS ,
	  KC_TRNS , KC_A , KC_S , KC_D    , KC_F    , KC_G     , KC_H    , KC_J    , KC_K    , KC_L   , KC_SCLN , KC_TRNS ,
	  KC_TRNS , KC_Z , KC_X , KC_C    , KC_V    , KC_B     , KC_N    , KC_M    , KC_COMM , KC_DOT , KC_SLSH , KC_TRNS ,
							  KC_TRNS , KC_LALT , KC_SPACE , MO(ARR) , KC_TRNS , KC_TRNS
```

fauxpark commented on keyboards/crkbd/keymaps/ollyhayes/keymap.c:
```diff
@@ -0,0 +1,112 @@
+//  +---------+---------------+------+--------+---------+-----------+-----------+------+--------+------+---------+---------+
+  ),
+
+  [FUNCTIONS] = LAYOUT_split_3x6_3( \
+//  +---------------+---------+---------+---------+---------+---------+---------+---------+---------+-------+---------+---------+
+	  KC_TRNS       , KC_TRNS , KC_TRNS , KC_F8   , KC_F9   , KC_F12  , KC_F12  , KC_F7   , KC_F8   , KC_F9 , KC_TRNS , KC_TRNS , \
+	  TG(FUNCTIONS) , KC_TRNS , KC_F4   , KC_F5   , KC_F6   , KC_F11  , KC_F11  , KC_F4   , KC_F5   , KC_F6 , KC_TRNS , KC_TRNS , \
+	  KC_TRNS       , KC_TRNS , KC_F1   , KC_F2   , KC_F3   , KC_F10  , KC_F10  , KC_F1   , KC_F2   , KC_F3 , KC_TRNS , KC_TRNS , \
+										  KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS                               \
```
```suggestion
  [FUNCTIONS] = LAYOUT_split_3x6_3(
//  +---------------+---------+---------+---------+---------+---------+---------+---------+---------+-------+---------+---------+
	  KC_TRNS       , KC_TRNS , KC_TRNS , KC_F8   , KC_F9   , KC_F12  , KC_F12  , KC_F7   , KC_F8   , KC_F9 , KC_TRNS , KC_TRNS ,
	  TG(FUNCTIONS) , KC_TRNS , KC_F4   , KC_F5   , KC_F6   , KC_F11  , KC_F11  , KC_F4   , KC_F5   , KC_F6 , KC_TRNS , KC_TRNS ,
	  KC_TRNS       , KC_TRNS , KC_F1   , KC_F2   , KC_F3   , KC_F10  , KC_F10  , KC_F1   , KC_F2   , KC_F3 , KC_TRNS , KC_TRNS ,
										  KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS
```

ollyhayes commented:
I've rebased onto the latest master and squashed it all into one commit, hope that's okay!

I think I've address all the comments, is there anything else I should do before this could be merged?

Thanks for all your help :slightly_smiling_face:
